llmfoundry/data/packing.py

```diff
@@ -224,50 +235,141 @@ def first_fit_bin_packing(
     return packed_examples[:num_bins], sum(
         bin_sizes[:num_bins]), sum(sizes), sorted_bins[num_bins:]
 
+def auto_packing_ratio(dataloader_cfg: DictConfig,
+                       tokenizer: PreTrainedTokenizerBase,
+                       device_batch_size: int):
+    """Find a packing ratio that minimizes padding with zero waste.
+
+    Args:
+        dataloader_cfg (DictConfig): The dataloader configuration for profiling.
+        tokenizer (PreTrainedTokenizerBase): The tokenizer for profiling.
+        device_batch_size (int): The size of the batches (number of examples) per device.
+
+    Returns:
+        A packing ratio that minimizes padding while maintaining zero waste.
+    """
+    # min_ratio = 2
+    # max_ratio = 2
+    # num_packing_ratios = 1
+    # profiling_results = profile_packing(dataloader_cfg, tokenizer, min_ratio,
+    #                                     max_ratio, num_packing_ratios,
+    #                                     device_batch_size)
+
+    # # Obtain the maximum packing_ratio/minimum padding that has no waste.
+    # i = 0
+    # waste = 0
+    # packing_ratio = 1
+    # while i < len(profiling_results) and waste == 0:
+    #     packing_ratio, _, waste = profiling_results[i]
+    #     i += 1
```

Comment:
uncomment and update min/max ratios as appropriate

probably gonna go for something like max_ratio = max_seq_len / 100
and num_packing_ratios = 15